### PR TITLE
FinNexus Adapter

### PR DESCRIPTION
**Pull request by Jeffqg123**
FinNexus Adapter is implemented , deployed on thereum mainnet ,tested and verified

FinNexusAssetAdapter:	0xd802469c0f5b7f9190289a977af0a7838eb8ae3a
FinNexusDebtAdapter:	0xdd9dbbd8e99c5ab4aaee985d3c7f13568f61517e
FinNexusTokenAdapter：	0xa884bb54d511ace355a924ac996397fb524cf35f

### CODE REVIEW

**Comment by sobolev-igor** [comment]
Hi @Jeffqg123! 

Could you please add tests into the repo as well as TokenAdapter for "FPT token", so it could be possible to fetch how many underlying tokens can user get for 1e18 of FPT token.

**Comment by Jeffqg123** [comment]
> Hi @Jeffqg123!
> 
> Could you please add tests into the repo as well as TokenAdapter for "FPT token", so it could be possible to fetch how many underlying tokens can user get for 1e18 of FPT token.

Hi @Igor Sobolev
The test case for fnx adapter is added, please review it

Thanks

File: contracts/adapters/finnexus/FinNexusDebtAdapter.sol [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    address internal constant FNX_ETH = 0x0000000000000000000000000000000000000000;
+    address internal constant USDC= 0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48;
+    address internal constant ETH = 0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE;
+
+    /**
+     * @return Amount of FNX/USDC/FNX locked on the protocol by the given account.
+     * @dev Implementation of ProtocolAdapter interface function.
+     */
+    function getBalance(address token, address account) external view override returns (uint256) {
+        if (token==FNX || token==ETH || token==USDC) {
```
Comment (sobolev-igor) [inline contracts/adapters/finnexus/FinNexusDebtAdapter.sol:51]:
```suggestion
        if (token == FNX || token == ETH || token == USDC) {
```

File: contracts/adapters/finnexus/FinNexusDebtAdapter.sol [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+ * @title Asset adapter for FinNexus option protocol.
+ * @dev Implementation of ProtocolAdapter interface.
+ * @author jeffqg123 <forestjqg@163.com>
+ */
+contract FinNexusDebtAdapter is ProtocolAdapter {
+
+    string public constant override adapterType = "Debt";
+    string public constant override tokenType = "ERC20";
+
+    address public constant CollecteralPool = 0x20081862B301C1ef9357BC0a125Aa414db210f2c;
```
Comment (sobolev-igor) [inline contracts/adapters/finnexus/FinNexusDebtAdapter.sol:40]:
```suggestion
    address public constant COLLATERAL_POOL = 0x20081862B301C1ef9357BC0a125Aa414db210f2c;
```

File: contracts/adapters/finnexus/FinNexusDebtAdapter.sol [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    address internal constant USDC= 0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48;
+    address internal constant ETH = 0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE;
+
+    /**
+     * @return Amount of FNX/USDC/FNX locked on the protocol by the given account.
+     * @dev Implementation of ProtocolAdapter interface function.
+     */
+    function getBalance(address token, address account) external view override returns (uint256) {
+        if (token==FNX || token==ETH || token==USDC) {
+            return FinNexusCollecteralPool(CollecteralPool).getUserInputCollateral(account,token==ETH?FNX_ETH:token);
```
Comment (sobolev-igor) [inline contracts/adapters/finnexus/FinNexusDebtAdapter.sol:52]:
```suggestion
            return FinNexusCollecteralPool(COLLATERAL_POOL).getUserInputCollateral(
                account,
                token == ETH ? FNX_ETH : token
            );
```

File: contracts/adapters/finnexus/FinNexusTokenAdapter.sol [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+            if(i==2) {
+                tokenPrice = 1e6 * tokenPrice;
+            } else {
+                tokenPrice = 1e18 * tokenPrice;
+            }
+            
+            underlyingTokens[i] = Component({
+                token: i==0?0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE:underlyingAddress[i],
+                tokenType: "ERC20",
+                rate: tokenPrice/fptWorth
```
Comment (sobolev-igor) [inline contracts/adapters/finnexus/FinNexusTokenAdapter.sol:89]:
```suggestion
                rate: tokenPrice / fptWorth
```

File: contracts/adapters/finnexus/FinNexusTokenAdapter.sol [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+            uint256 tokenPrice = FNXOracle(oracle).getPrice(underlyingAddress[i]);
+            
+            if(i==2) {
+                tokenPrice = 1e6 * tokenPrice;
+            } else {
+                tokenPrice = 1e18 * tokenPrice;
+            }
+            
+            underlyingTokens[i] = Component({
+                token: i==0?0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE:underlyingAddress[i],
```
Comment (sobolev-igor) [inline contracts/adapters/finnexus/FinNexusTokenAdapter.sol:87]:
```suggestion
                token: i == 0 ? 0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE : underlyingAddress[i],
```

File: contracts/adapters/finnexus/FinNexusTokenAdapter.sol [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+
+        for (uint256 i = 0; i < underlyingTokens.length; i++) {
+            uint256 fptWorth = OptionsManagerV2(optManager).getTokenNetworth();
+            uint256 tokenPrice = FNXOracle(oracle).getPrice(underlyingAddress[i]);
+            
+            if(i==2) {
+                tokenPrice = 1e6 * tokenPrice;
+            } else {
+                tokenPrice = 1e18 * tokenPrice;
+            }
```
Comment (sobolev-igor) [inline contracts/adapters/finnexus/FinNexusTokenAdapter.sol:84]:
```suggestion
            tokenPrice = (i == 2) ? tokenPrice * 1e6 : tokenPrice * 1e18;
```

File: contracts/adapters/finnexus/FinNexusTokenAdapter.sol [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+    /**
+     * @return Array of Component structs with underlying tokens rates for the given token.
+     * @dev Implementation of TokenAdapter interface function.
+     */
+    function getComponents(address) external view override returns (Component[] memory) {
+        
+        Component[] memory underlyingTokens = new Component[](underlyingAddress.length);
+
+        for (uint256 i = 0; i < underlyingTokens.length; i++) {
+            uint256 fptWorth = OptionsManagerV2(optManager).getTokenNetworth();
```
Comment (sobolev-igor) [inline contracts/adapters/finnexus/FinNexusTokenAdapter.sol:77]:
```suggestion

        uint256 fptWorth = OptionsManagerV2(optManager).getTokenNetworth();
        uint256 tokenPrice;
        for (uint256 i = 0; i < underlyingTokens.length; i++) {
```

File: contracts/adapters/finnexus/FinNexusTokenAdapter.sol [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+            uint256 tokenPrice = FNXOracle(oracle).getPrice(underlyingAddress[i]);
```
Comment (sobolev-igor) [inline contracts/adapters/finnexus/FinNexusTokenAdapter.sol:78]:
```suggestion
           tokenPrice = FNXOracle(ORACLE).getPrice(underlyingAddress[i]);
```

File: contracts/adapters/finnexus/FinNexusTokenAdapter.sol [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+ * @author jeffqg123 <forestjqg@163.com>
+ */
+contract FinNexusTokenAdapter is TokenAdapter {
+
+    address public  constant optManager = 0xfa30ec96De9840A611FcB64e7312f97bdE6e155A;
+    address public  constant oracle = 0x940b491905529542Ba3b56244A06B1EBE11e71F2;
+
+    address[] internal underlyingAddress = [0x0000000000000000000000000000000000000000,0xeF9Cd7882c067686691B6fF49e650b43AFBBCC6B,0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48];
+                                                           //  ,
+                                                           //  ];
```
Comment (sobolev-igor) [inline contracts/adapters/finnexus/FinNexusTokenAdapter.sol:54]:
```suggestion
    address public constant OPT_MANAGER = 0xfa30ec96De9840A611FcB64e7312f97bdE6e155A;
    address public constant ORACLE = 0x940b491905529542Ba3b56244A06B1EBE11e71F2;

    address[] internal underlyingAddress =[
        0x0000000000000000000000000000000000000000,
        0xeF9Cd7882c067686691B6fF49e650b43AFBBCC6B,
        0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48
    ];
```

Comment (sobolev-igor) [reply contracts/adapters/finnexus/FinNexusTokenAdapter.sol:54]: I'd recommend you to use constants, as well as `ETH_ADDRESS` for `0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE`

**Review by sobolev-igor — CHANGES_REQUESTED** [review]

**Comment by Jeffqg123** [comment]
Hi @igor Sobolev
Thank you for your midification.
Then what should I do for next step?
Thanks

File: contracts/adapters/finnexus/FinNexusTokenAdapter.sol [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+
+    /**
+     * @return Array of Component structs with underlying tokens rates for the given token.
+     * @dev Implementation of TokenAdapter interface function.
+     */
+    function getComponents(address) external view override returns (Component[] memory) {
+        
+        Component[] memory underlyingTokens = new Component[](underlyingAddress.length);
+
+        uint256 fptWorth = OptionsManagerV2(OPT_MANAGER).getTokenNetworth();
```
Comment (sobolev-igor) [inline contracts/adapters/finnexus/FinNexusTokenAdapter.sol:78]: is the case of `fptWorth == 0` worth considering or it's impossible?

**Comment by sobolev-igor** [comment]
Now, I'll ask you to add some protocol metadata:
1. Name (FinNexus, I suppose)
2. 1-line description
3. icon (usually we use 200x200 icons)
4. website url

File: contracts/adapters/finnexus/FinNexusTokenAdapter.sol (hunk 10 above)
Comment (Jeffqg123) [reply contracts/adapters/finnexus/FinNexusTokenAdapter.sol:78]: fptWorth is always bigger than 0,

**Comment by Jeffqg123** [comment]
> metadata

where should I put these protocol metadata?

**Comment by Jeffqg123** [comment]
> > metadata
> 
> where should I put or upload these protocol metadata?

**Comment by Jeffqg123** [comment]
Name:   FinNexus


1-line description:  FinNexus builds universal decentralized protocols for options with pooled liquidity, which works under the Multi-Asset Single-Pool mechanism.

icon (usually we use 200x200 icons): 

website url: https://www.finnexus.io

**Comment by Jeffqg123** [comment]
If have any problem for metadata,please tell me.

Thanks

**Comment by sobolev-igor** [comment]
Cool, everything is fine, I'll add it asap

**Pull request merged**